arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.cpp

```diff
@@ -215,4 +219,27 @@ auto DocumentStateShardHandler::lockShard(ShardID const& shard,
   return trx;
 }
 
+auto DocumentStateShardHandler::prepareShardsForLogReplay() noexcept -> void {
+  for (auto const& shard : getAvailableShards()) {
+    // The inverted indexes cannot work with duplicate LocalDocumentIDs within the same
+    // commit interval. They however can if we have a commit in between the two.
+    // If we replay one log we know there can never be a duplicate LocalDocumentID.
+    for (auto const& index : shard->getIndexes()) {
+      {
+        auto maybeInvertedIndex =
+            std::dynamic_pointer_cast<iresearch::IResearchInvertedIndex>(index);
+        if (maybeInvertedIndex) {
```

Comment:
ArangoDB indexes has type

Reply:
Added ifs for the type.

Reply:
Then we don't need to check cast result. More to say we have a more robust machinery to do this: 

``` cpp
    if (index->type() == Index::TRI_IDX_TYPE_INVERTED_INDEX) {
       basics::downCast<IResearchRocksDBInvertedIndex>(*index).commit(true);
    } else if (index->type() == Index::TRI_IDX_TYPE_IRESEARCH_LINK) {
       basics::downCast<IResearchRocksDBLink>(*index).commit(true);
    }
```
It does the same and has all necessary asserts inside.

Reply:
Perfect, exactly what I was looking for :+1: